traitlets/config/configurable.py

```diff
@@ -32,9 +33,47 @@ class MultipleInstanceError(ConfigurableError):
 
 class Configurable(HasTraits):
 
+    envvar_eval_context = {}
     config = Instance(Config, (), {})
     parent = Instance('traitlets.config.configurable.Configurable', allow_none=True)
 
+    @classmethod
+    def setup_class(cls, classdict):
+        for k, v in iteritems(classdict):
+            if isinstance(v, BaseDescriptor):
+                v.class_init(cls, k)
+                if isinstance(v, TraitType) and 'envvar' in v.metadata:
+                    v.default_value = cls._default_from_envvar_internal(v)
+
+    @classmethod
+    def _default_from_envvar_internal(cls, trait):
+        """Generate a default for the trait from os.environ
+
+        The variable strings are converted to values using ``eval`` where
+        the the context is defined by ``cls.envvar_eval_context``.
+        """
+        try:
+            # must have 'envvar' in metadata
+            vname = trait.metadata['envvar']
+            string = os.environ[vname]
+        except KeyError:
+            raise TraitError(
+                "{type} trait with name {name} expected "
+                "os.environ[{envvar!r}] to be set.".format(
+                    type=type(trait).__name__,
+                    envvar=vname,
+                    # repring here to add quotes in the formatted error message.
+                    name=repr(trait.name) if trait.name else "<anonymous>",
+                )
+            )
+        try:
+            value = eval(string, {}, cls.envvar_eval_context)
```

Comment:
It seems like the string=> value conversion should depend on the trait type, and not use eval. `from_env` should probably be a method on each TraitType that supports coming from an env (e.g. only Unicode, numbers).